### ISSUE TITLE
Wiki parser not recognizing ")" as part of a link

### ISSUE THREAD

**Issue by zwn** [zim-desktop-wiki/zim-desktop-wiki#219]
The current wiki parser does not recognize url "https://en.wikipedia.org/wiki/80/20_(framing_system)" - it does not include the ending ")".

**Comment by jaap-karssenberg** [comment]
This is intentional to allow writing e.g. `(http://example.com)` open to improve the rule for this

Interestingly, observe how github markdown handles these:
- Not part of link: (http://example.com)
- Part of link: http://example.com()
- Also not part of link: (See http://example.com)
- Even not here http://example.com)

Seems the rule here is that the ")" is only included if the link also contains the "("

**Comment by zwn** [comment]
What probably surprised me the most was that I have found no way to create such a link. I have noticed the link being wrong, used the GUI to fix it, only to find out it is still broken.

Can the text be enclosed somehow to help it parse as a link?

**Comment by jaap-karssenberg** [comment]
Afraid not. Enclosing in "[[...]]" in the source works, but they disappear when editing the page again

**Comment by jaap-karssenberg** [comment]
See https://github.github.com/gfm/#autolinks-extension- for a specification on how to handle closing ")" at the end of an URL better

**Comment by myrdd** [comment]
I'm running into this quite often, mostly for links to wikipedia. My workaround was to append `#top` at the end of the link, like https://en.wikipedia.org/wiki/80/20_(framing_system)#top, which works — as long as you do not want to link to a _section_ ending itself with a closing parenthesis (`)`), like https://de.wikipedia.org/wiki/Variable_(Programmierung)#Sichtbarkeitsbereich_von_Variablen_(Scope).

**Comment by myrdd** [comment]
similar issue with `=---` at the end (bugzilla), e.g.

> `https://bugs.kde.org/buglist.cgi?resolution=---`

is parsed/rendered as

> <code>[https://bugs.kde.org/buglist.cgi?resolution](https://bugs.kde.org/buglist.cgi?resolution)=---</code>
> <code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;^^^^__ not part of the link</code>

The URL ”fragment“ (`#...`) I append in the case of bugzilla is `#bugzilla-body` — e.g. `https://bugs.kde.org/buglist.cgi?resolution=---#bugzilla-body` — which is parsed/rendered correctly. Or simply `#foo` ;)

**Comment by myrdd** [comment]
Just another example:

> `https://vimhelp.org/options.txt.html#'iskeyword'`

Parsed by Zim like this:

> <code>[https://vimhelp.org/options.txt.html](https://vimhelp.org/options.txt.html)#'iskeyword'</code>

**Comment by jaap-karssenberg** [comment]
Fix by implementing rules as described here https://github.github.com/gfm/#autolinks-extension-

**Comment by myrdd** [comment]
**tl;dr:** enclosing an URL in `[[` and `]]` fixes almost all external link parsing issues, but Zim removes the brackets and therefore breaks the link.

---

@jaap-karssenberg I can see you've linked the related issues #400 (apostrophe) and #421 (dash) here. Please note that not all possible cases are covered by the Github Flavoured Markdown (GFM). Examples:

* `'` at the end
* `)` at the end w/o opening parenthesis
* `]` at the end w/o opening bracket (compare #596)

It would be nice if any URL inserted via ”Insert Link“ or ”Edit Link“ dialog was written to disk in a format that was parsed correctly after reloading.

Currently, as far as I've observed, all external links *not* having a custom ”Text“ (e.g. `[[http://test/|Custom Text]]`) are written to disk *without* any brackets (`[[` + `]]`).

So I suggest the following: (scenario + expected behavior)

1. the user opens an ”Insert Link“ or ”Edit Link“ dialog
    * no custom ”Text“, it's the same as the URL
2. the user clicks on ”Link“ (press enter)
3. Just for checking purposes, Zim parses the URL as if it would have come from a Zim `.txt` file and **compares** the parsing result with the raw URL.
    * if it's the same, the URL can safely stored as-is to disk
    * if it's different, then the URL needs to be escaped (i.e. `%xx`) or wrapped inside double brackets (`[[` + `]]`).

I could verify that the following *raw* Zim `.txt` lines are parsed correctly in Zim v0.71.0:

```
apostrophe:
[[https://graphemica.com/']]
https://graphemica.com/%27

closing parens
[[https://graphemica.com/)]]
https://graphemica.com/%29

closing bracket:
https://graphemica.com/%5D
```

However, the `[[...]]` variant is changed by Zim itself back to `...` and thus breaks the link.

**Comment by seihtam** [comment]
This is not totally fixed in 0.73.2. Try to paste these links into Zim:
https://payatu.com/blog/asmita-jha/iot-security---part-17-(101---hardware-attack-surface:-uart)
https://payatu.com/blog/shakir/iot-security-part-14-(introduction-to-and-identification-of-hardware-debug-ports)

The closing parenthesis will not be recognized as part of the url.

But others like this one works as expected:
https://payatu.com/blog/shakir/iot-security-part-13-(introduction-to-hardware-recon)

**Comment by jaap-karssenberg** [comment]
Looks like an issue specific with copy-paste -- after reloading the page
(Control-R) it is recognized correctly.

-- Jaap

**Comment by jaap-karssenberg** [comment]
Need to implement new logic to recognize url in "plain" text parser - copy
from "wiki" parser.

-- Jaap
